plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/AnalyticsPingManager.java

```diff
@@ -133,14 +133,30 @@ void registerOptInStatus(boolean optedIn) {
    * If the user has never seen the opt-in dialog or set the opt-in preference beforehand,
    * this method can potentially present an opt-in dialog at the top workbench level. If you
    * are calling this method inside another modal dialog, consider using {@link #sendPing(
-   * String, String, String, Shell)} and pass the {@Shell} of currently open modal dialog.
+   * String, String, String, Shell)} and pass the {@Shell} of the currently open modal dialog.
    * (Otherwise, the opt-in dialog won't be able to get input until the workbench can get input.)
+   *
+   * Safe to call from non-UI contexts.
    */
   public void sendPing(String eventName, String metadataKey, String metadataValue) {
     sendPing(eventName, metadataKey, metadataValue, null);
   }
 
-  public void sendPing(String eventName, String metadataKey, String metadataValue,
+  public void sendPing(final String eventName,
+      final String metadataKey, final String metadataValue, final Shell parentShell) {
+    if (Display.getCurrent() != null) {  // Check if we are in the UI thread.
```

Comment:
What do you think? Currently, `sendPing()` may potentially show the opt-in dialog (if a user has never seen it). That's the only reason we want the UI thread. I had thought about a few alternatives before:

1. For headless jobs (e.g., local run), add something like `sendPingUiFree()` which never pops up dialog.
2. Forget about UI vs. non-UI inside `AnalyticsPingManager`, and have callers responsible for wrapping `sendPing()` with `Display#asyncExec()` if calling from headless jobs.
3. Check if we are in the UI thread, and if not, never pops up the dialog.

The approach in this PR is to liberate any concern from callers about UI vs. non-UI and always be able to show the dialog if necessary. Suggestions?